SparseGrids/tsgLinearSolvers.cpp

```diff
@@ -237,6 +256,109 @@ void TasmanianTridiagonalSolver::decompose(int n, std::vector<double> &d, std::v
+    }
+    norm = std::max(norm, std::fabs(diag[n]) + std::fabs(off_diag[n-1]));
+    weights[n-1] = 0.0;
+    weights[0] = 1.0; // Fix the bug in the ALGOL code.
+    double eps = norm * std::numeric_limits<double>::epsilon();
+    size_t m = n;
+    double lambda{norm}, lambda1{norm}, lambda2{norm}, rho{norm};
+
+    // INSPECT block from ALGOL code.
+    // ALGOL COMMENT: Look for convergence of lower diagonal element.
+    while (m > 0) {
+        if (std::fabs(off_diag[m-1]) <= eps) {
+            nodes[m-1] = diag[m];
+            weights[m-1] = mu0 * weights[m-1] * weights[m-1];
+            rho = lambda1 < lambda1 ? lambda1 : lambda2;
+            m = m-1;
+            continue;
+        }
+        // ALGOL COMMENT: Small off diagonal element means matrix can be split.
+        int k = m-1;
+        while (std::fabs(off_diag[k-1]) > eps) k--;
+        // ALGOL COMMENT: Find eigenvalues of lower 2-by-2 and select accelerating shift.
+        double b2 = off_diag[m-1] * off_diag[m-1];
+        double det = std::sqrt((diag[m-1] - diag[m]) * (diag[m-1] - diag[m]) + 4.0 * b2);
+        double aa = diag[m-1] + diag[m];
+        lambda2 = 0.5 * (aa >= 0 ? aa + det : aa - det);
+        lambda1 = (diag[m-1] * diag[m] - b2) / lambda2;
+        double eigmax = std::max(lambda1, lambda2);
+        if (std::fabs(eigmax - rho) <= 0.125 * std::fabs(eigmax)) {
+            lambda = eigmax;
+        }
+        rho = eigmax;
+        // ALGOL COMMENT: Transform block from k to m.
+        double cj = off_diag[k];
+        off_diag[k-1] = diag[k] - lambda;
+        for (size_t j=k; j<=m-1; j++) {
+            double r = std::sqrt(cj * cj + off_diag[j-1] * off_diag[j-1]);
+            double st = cj / r;            double st2 = st * st;
+            double ct = off_diag[j-1] / r; double ct2 = ct * ct;
+            double sc = st * ct;           double aj = diag[j];
+            double bj = off_diag[j];       double wj = weights[j-1];
+            // Order below is important!
+            diag[j] = aj * ct2 + 2.0 * bj * sc + diag[j+1] * st2;
+            off_diag[j] = (aj - diag[j+1]) * sc + bj * (st2 - ct2);
+            diag[j+1] = aj * st2 - 2.0 * bj * sc + diag[j+1] * ct2;
+            cj = off_diag[j+1] * st;
+            off_diag[j+1] = -off_diag[j+1] * ct;
+            off_diag[j-1] = r;
+            // Account for the offset of the indices.
+            weights[j-1] = wj * ct + weights[j] * st;
+            weights[j] = wj * st - weights[j] * ct;
+        }
+        off_diag[k-1] = 0.0;
+    }
+
+    // SORT block from ALGOL code.
+    // ALGOL COMMENT: Arrange abscissas in ascending order.
+    // NOTE: the original code used an exchange sort, which is O(n^2).
+    std::vector<size_t> I(n);
+    for (size_t i=0; i<n; i++) I[i] = i;
+    std::sort(I.begin(), I.end(), [&nodes](size_t i, size_t j){return nodes[i] < nodes[j];});
+    for (size_t i=0; i<n; i++) {
+        if (i != I[i]) {
+            double tmp_node, tmp_weight;
```

Comment:
Be careful when you read old code and old papers, especially in the 80s.

Right here, why declare the variables and only the initialize them?
Also this whole section seems it can be done with just a few swap operations suing `std::swap()`

Reply:
This sort block is actually my own implementation of sorting in-place (the original ALGOL code uses an O(n^2) exchange sort instead). Let me see if I can clean it up a little bit.

Reply:
Okay, I pushed a version that only uses only three swaps in the main while-loop. This one should be equivalent to the previous (longer) version.